### PR TITLE
[All] Base strategy creation 

### PR DESCRIPTION
**Pull request by rubenzorrilla**
**Description**
As we already discussed when merging the new explicit strategies, in this PR I'm adding the `ImplicitSolvingStrategy` class which is identical to the previous `SolvingStrategy` one. This allows deriving both the new `ImplicitSolvingStrategy` and `ExplicitStrategy` from a new base `SolvingStrategy` class, which has no linear solver template argument. By doing this we ensure that there is a unique API definition for all the strategies, which can be found in `SolvingStrategy`.

I took the chance to add a new `GetMoveMeshFlag()` method which is intended to be consistent why the other `Get...` methods of the strategy. This will allow to deprecate the current `MoveMeshFlag()` method. 

Besides that, I'd like to hear your opinion about having the `mRebuildLevel` in the derived implicit/explicit strategies rather than in the base `SolvingStrategy` one. The point is that the rebuild level in explicit strategies is either true or false, and could eventually be removed since the move mesh flag already indicates if the mass vectors need to be recomputed. 

**Changelog**
- Creation of the `ImplicitSolvingStrategy` class from the previous `SolvingStrategy` one
- Creation of a new `SolvingStrategy` class from which the `ImplicitSolvingStrategy` and `ExplicitSolvingStrategy` derive

### CODE REVIEW

File: kratos/solving_strategies/strategies/solving_strategy.h [hunk 1]
```diff
@@ -160,10 +154,9 @@ class SolvingStrategy
      * @param rModelPart The model part of the problem
      * @param ThisParameters The configuration parameters
      */
-    virtual typename ClassType::Pointer Create(
+    typename ClassType::Pointer Create(
```
Comment (philbucher) [inline kratos/solving_strategies/strategies/solving_strategy.h]: I think this needs to be virtual

File: kratos/solving_strategies/strategies/solving_strategy.h [hunk 2]
```diff
@@ -414,51 +388,11 @@ class SolvingStrategy
...
-        KRATOS_ERROR << "GetSystemVector not implemented in base SolvingStrategy" << std::endl;
-
-        KRATOS_CATCH("");
-    }
-
-    /**
-     * @brief This method returns the solution vector
-     * @return The Dx vector
-     */
-    virtual TSystemVectorType& GetSolutionVector()
```
Comment (philbucher) [inline kratos/solving_strategies/strategies/solving_strategy.h]: this was only recently added: #6763
I think it cannot be removed

Comment (rubenzorrilla) [reply kratos/solving_strategies/strategies/solving_strategy.h]: It is added in the implicit strategy. It makes no sense to have it in the base one as it's meaningless in the explicit one.

File: kratos/solving_strategies/strategies/solving_strategy.h (hunk 1 above)
Comment (rubenzorrilla) [reply kratos/solving_strategies/strategies/solving_strategy.h]: I did it virtual in the very first stage but it didn't work. I think that the problem is that the implicit strategy has different number of template arguments, then we cannot point with a `SolvingStrategy::Pointer` to an `ImplicitSolvingStrategy`.

File: kratos/solving_strategies/strategies/solving_strategy.h (hunk 2 above)
Comment (philbucher) [reply kratos/solving_strategies/strategies/solving_strategy.h]: I know, but the Factories will give most probably the baseclasses
and if those fcts are missing in the baseclass they cannot be accesses

I agree that they ake no sense in explicit but sometimes it is necessary (e.g. the B&S baseclass also has some fcts that don't make sene for Block or ElimB&S)

File: kratos/solving_strategies/strategies/solving_strategy.h (hunk 1 above)
Comment (pooyan-dadvand) [reply kratos/solving_strategies/strategies/solving_strategy.h]: So you may have an empty linear solver passed to the explicit solver by default

File: kratos/solving_strategies/strategies/solving_strategy.h (hunk 2 above)
Comment (rubenzorrilla) [reply kratos/solving_strategies/strategies/solving_strategy.h]: Agree and done. When doing this I didn't think about the factories. Thanks for pointing it out.

File: kratos/solving_strategies/strategies/solving_strategy.h (hunk 1 above)
Comment (rubenzorrilla) [reply kratos/solving_strategies/strategies/solving_strategy.h]: Done. Finally the solution was as simple as defining an extra type for the base class :facepalm:

**Comment by rubenzorrilla** [comment]
I think that this is now ready for review. The `Create` return type issue has been solved and also the base class method addition towards the future creation of factories.

**Comment by loumalouomega** [comment]
> I think that this is now ready for review. The `Create` return type issue has been solved and also the base class method addition towards the future creation of factories.

I thought I added the Create in the factories PR

**Comment by rubenzorrilla** [comment]
> > I think that this is now ready for review. The `Create` return type issue has been solved and also the base class method addition towards the future creation of factories.
> 
> I thought I added the Create in the factories PR

I was referring to the issue pointed by @philbucher. The `Create` was not virtual in the first draft version of the PR. Now should be fine. `Create` is `virtual` and returns a pointer to the base `SolvingStrategy` class.

**Comment by rubenzorrilla** [comment]
Considering that this is a large conflict prone PR, it would be nice to merge it as soon as possible so any feedback in this regard would be appreciated.

**Review by loumalouomega — DISMISSED** [review]

File: kratos/templates/template_application/custom_python/add_custom_strategies_to_python.cpp.in [hunk 3]
```diff
@@ -42,7 +42,7 @@ void  AddCustomStrategiesToPython(pybind11::module& m)
     typedef UblasSpace<double, Matrix, Vector> LocalSpaceType;
 
     typedef LinearSolver<SparseSpaceType, LocalSpaceType > LinearSolverType;
-    typedef SolvingStrategy< SparseSpaceType, LocalSpaceType, LinearSolverType > BaseSolvingStrategyType;
+    typedef ImplicitSolvingStrategy< SparseSpaceType, LocalSpaceType, LinearSolverType > BaseSolvingStrategyType;
```
Comment (philbucher) [inline kratos/templates/template_application/custom_python/add_custom_strategies_to_python.cpp.in:45]: hm guess this should be changed

can you add the second exposure (with name `ImplicitSolvingStrategy`) and deprecate the other one?
For me fine to do it in a separate PR to merge this one soon

File: kratos/solving_strategies/strategies/solving_strategy.h [hunk 4]
```diff
@@ -356,11 +349,19 @@ class SolvingStrategy
...
+        return *mpModelPart;
+    };
+
+    /**
+     * @brief Operations to get the pointer to the model
+     * @return *mpModelPart: The model part member variable
+     */
+    const ModelPart& GetModelPart() const
     {
-        KRATOS_ERROR_IF_NOT(mpModelPart) << "ModelPart in the SolvingStrategy is not initialized" << std::endl;
```
Comment (philbucher) [inline kratos/solving_strategies/strategies/solving_strategy.h:363]: I think it is important to leave this error!

File: kratos/solving_strategies/strategies/solving_strategy.h [hunk 5]
```diff
@@ -356,11 +349,19 @@ class SolvingStrategy
 
     /**
      * @brief Operations to get the pointer to the model
-     * @return mpModelPart: The model part member variable
+     * @return *mpModelPart: The model part member variable
      */
-    inline ModelPart& GetModelPart()
+    ModelPart& GetModelPart()
+    {
+        return *mpModelPart;
```
Comment (philbucher) [inline kratos/solving_strategies/strategies/solving_strategy.h:356]: also it should be checked here that the pointer is valid, otherwise you get a segfault!

File: kratos/python_scripts/analysis_stage.py [hunk 6]
```diff
@@ -346,7 +346,7 @@ def __CreateListOfProcesses(self):
 
     def __CheckIfSolveSolutionStepReturnsAValue(self, is_converged):
         """In case the solver does not return the state of convergence
-        (same as the SolvingStrategy does) then issue ONCE a deprecation-warning
+        (same as the ImplicitSolvingStrategy does) then issue ONCE a deprecation-warning
```
Comment (philbucher) [inline kratos/python_scripts/analysis_stage.py]:
```suggestion
        (same as the SolvingStrategy does) then issue ONCE a deprecation-warning

```

this was correct as is

File: kratos/solving_strategies/strategies/implicit_solving_strategy.h [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+
+/* External includes */
+
+/* Project includes */
+#include "includes/define.h"
+#include "includes/model_part.h"
+#include "solving_strategy.h"
+#include "solving_strategies/schemes/scheme.h"
+#include "solving_strategies/builder_and_solvers/builder_and_solver.h"
+#include "includes/kratos_parameters.h"
```
Comment (philbucher) [inline kratos/solving_strategies/strategies/implicit_solving_strategy.h:27]: please only include what is actually used

File: kratos/mappers/mapper.h [hunk 8]
```diff
@@ -35,7 +35,7 @@ namespace Kratos
 ///@{
 
 /// Base Class for all Mappers
-/** This is the base class for every mapper. This is the equivalent to a Kratos-SolvingStrategy.
+/** This is the base class for every mapper. This is the equivalent to a Kratos-ImplicitSolvingStrategy.
```
Comment (philbucher) [inline kratos/mappers/mapper.h]:
```suggestion
/** This is the base class for every mapper. This is the equivalent to a Kratos-SolvingStrategy.
```

was correct as is

File: applications/TrilinosApplication/custom_python/add_trilinos_strategies_to_python.cpp [hunk 9]
```diff
@@ -114,29 +117,34 @@ void AddStrategies(pybind11::module& m)
...
+        .def("Clear", &TrilinosBaseSolvingStrategyType::Clear)
+        .def("Check", &TrilinosBaseSolvingStrategyType::Check)
+        .def("InitializeSolutionStep", &TrilinosBaseSolvingStrategyType::InitializeSolutionStep)
+        .def("FinalizeSolutionStep", &TrilinosBaseSolvingStrategyType::FinalizeSolutionStep)
+        .def("SolveSolutionStep", &TrilinosBaseSolvingStrategyType::SolveSolutionStep)
+        .def("GetModelPart", [](TrilinosBaseSolvingStrategyType& self) -> ModelPart& { return self.GetModelPart(); })
+        ;
+
+    // Implicit strategy base class
+    py::class_< TrilinosImplicitSolvingStrategyType, typename TrilinosImplicitSolvingStrategyType::Pointer, TrilinosBaseSolvingStrategyType >(m, "TrilinosImplicitSolvingStrategy")
```
Comment (philbucher) [inline applications/TrilinosApplication/custom_python/add_trilinos_strategies_to_python.cpp:140]: as other comment, please make it consistent with the serial one (future PR is good for me)

**Review by philbucher — DISMISSED** [review]
looks very good to me now, thx for the effort (and patience...) !
Feel free to merge and address the minor comment I made in a future PR :)

File: kratos/python_scripts/analysis_stage.py (hunk 6 above)
Comment (rubenzorrilla) [reply kratos/python_scripts/analysis_stage.py]: Agree. Thanks for spotting!

File: kratos/mappers/mapper.h (hunk 8 above)
Comment (rubenzorrilla) [reply kratos/mappers/mapper.h]: Agree. Thanks!

File: kratos/solving_strategies/strategies/solving_strategy.h (hunk 5 above)
Comment (rubenzorrilla) [reply kratos/solving_strategies/strategies/solving_strategy.h:356]: Yes. Indeed the current implementation does it. I can't understand why I did it in this way. Thanks!

Comment (rubenzorrilla) [reply kratos/solving_strategies/strategies/solving_strategy.h:356]: Done.

File: kratos/solving_strategies/strategies/solving_strategy.h (hunk 4 above)
Comment (rubenzorrilla) [reply kratos/solving_strategies/strategies/solving_strategy.h:363]: Done.

File: kratos/solving_strategies/strategies/implicit_solving_strategy.h (hunk 7 above)
Comment (rubenzorrilla) [reply kratos/solving_strategies/strategies/implicit_solving_strategy.h:27]: Including the base `solving_strategy.h` should be sufficient. Done.

**Comment by loumalouomega** [comment]
Just to mention, this is going to conflict with the strategies factories that are still pending it fix

File: kratos/templates/template_application/custom_python/add_custom_strategies_to_python.cpp.in (hunk 3 above)
Comment (rubenzorrilla) [reply kratos/templates/template_application/custom_python/add_custom_strategies_to_python.cpp.in:45]: Makes sense. I can change it in here.

Comment (rubenzorrilla) [reply kratos/templates/template_application/custom_python/add_custom_strategies_to_python.cpp.in:45]: Indeed, maybe it makes more sense that we change what is done below to add an example of implicit and explicit strategy export.

File: applications/TrilinosApplication/custom_python/add_trilinos_strategies_to_python.cpp (hunk 9 above)
Comment (rubenzorrilla) [reply applications/TrilinosApplication/custom_python/add_trilinos_strategies_to_python.cpp:140]: I think I don't get your point at all. I guess you mean having the very same exposure in serial and Trilinos classes. Am I wrong?

**Review by philbucher — APPROVED** [review]

**Comment by rubenzorrilla** [comment]
> Just to mention, this is going to conflict with the strategies factories that are still pending it fix

OK. Lets merge it then. The minors can be addressed in a future PR.

**Comment by philbucher** [comment]
the nightly build fails bcs of #9004

**Comment by rubenzorrilla** [comment]
> the nightly build fails bcs of #9004

I'd go ahead then.

File: applications/TrilinosApplication/custom_python/add_trilinos_strategies_to_python.cpp (hunk 9 above)
Comment (philbucher) [reply applications/TrilinosApplication/custom_python/add_trilinos_strategies_to_python.cpp:140]: I only mean to expose it consistently:

~~~py
SolvingStrategy
ImplicitSolvingStrategy

TrilinosSolvingStrategy
TrilinosImplicitSolvingStrategy
~~~
